File: formula-cs/Formula/TokenTree/Node.cs
namespace Formula.TokenTree;

public abstract class Node<T>
{
    protected List<Node<T>>? Children;

    public Node<T> Add(Node<T> child) {
        Children ??= new List<Node<T>>();
        foreach (var existingChild in Children) {
            if (existingChild.Equals(child)) {
                return existingChild;
            }
        }
        Children.Add(child);
        return child;
    }

    public abstract void Walk(string text, int startIndex, int currentIndex, List<TokenMatch<T>> matches);

    protected void WalkChildren(string text, int startIndex, int currentIndex, List<TokenMatch<T>> matches) {
        if (Children == null)
        {
            return;
        }

        foreach (var child in Children)
        {
            child.Walk(text, startIndex, currentIndex, matches);
        }
    }
}

File: java/formula-java/src/main/java/org/formula/parse/tree/Node.java
package org.formula.parse.tree;

import java.util.ArrayList;
import java.util.List;

public abstract class Node<T> {
    protected List<Node<T>> children;

    public Node<T> add(Node<T> child) {
        if (children == null) {
            children = new ArrayList<>();
        }
        for (Node<T> existingChild : children) {
            if (existingChild.equals(child)) {
                return existingChild;
            }
        }
        children.add(child);
        return child;
    }

    public abstract void walk(String text, int startIndex, int currentIndex, List<TokenMatch<T>> matches);

    protected void walkChildren(String text, int startIndex, int currentIndex, List<TokenMatch<T>> matches) {
        if (children == null) {
            return;
        }

        children.forEach(child -> child.walk(text, startIndex, currentIndex, matches));
    }
}